src/Compilers/CSharp/Portable/Compiler/MethodCompiler.cs

```diff
@@ -1024,6 +1024,25 @@ public override object VisitField(FieldSymbol symbol, TypeCompilationState argum
                                 DefiniteAssignmentPass.Analyze(_compilation, methodSymbol, analyzedInitializers, diagsForCurrentMethod, requireOutParamsAssigned: false);
                             }
 
+                            foreach (var parameter in methodSymbol.Parameters)
+                            {
+                                bool isParameterUsed = false;
+                                foreach (var statement in analyzedInitializers.Statements)
+                                {
+                                    if (statement is BoundExpressionStatement boundExpression &&
+                                        boundExpression.Expression is BoundAssignmentOperator boundAssignment &&
+                                        boundAssignment.Right.ExpressionSymbol == parameter)
+                                    {
+                                        isParameterUsed = true;
+                                        break;
+                                    }
+                                }
+                                if (!isParameterUsed)
+                                {
+                                    _diagnostics.Add(ErrorCode.WRN_UnusedRecordCtorParameter, parameter.Locations[0], parameter.Name);
```

Comment:
I think I need to put an if condition to make sure this new chunk of code is only executed for record synthesized constructors, right?

Reply:
The main problem with this code is that it only does a shallow analysis. Consider `int Property { get; init; } = StaticMethod(parameter);`. The parameter is used, but your analysis would miss this. That is was data flow analysis does (it walks the tree deeply).

Reply:
Yeah the approach I used misses such cases :(

The flow analysis seem to be already done in:

https://github.com/dotnet/roslyn/blob/b07cd9d710b3933d218a732b90287553ab0b8c72/src/Compilers/CSharp/Portable/Compiler/MethodCompiler.cs#L1024

But I'm not sure how to update it to report diagnostic for unused record positional parameter.

If the fix is going to be complex, I'd probably close the PR and let someone more experienced do it.

Probably it's just a small check for `_usedVariables` somewhere, but I can't spot where the check should go.